File: tfcc_runtime/operations/base/transpose.cpp
// Copyright 2021 Wechat Group, Tencent
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "transpose.h"

#include <algorithm>

#include "tfcc_runtime/framework/graph.h"
#include "tfcc_runtime/proto/operations/base.pb.h"
#include "tfcc_runtime/utils/commutils.h"

namespace tfcc {
namespace runtime {
namespace base {

template <class T>
static const char* _wrapper_transpose(
    const tfcc::Tensor<T>* a, const std::vector<size_t>* perm, tfcc::Variable<T>* result) noexcept {
  try {
    *result = tfcc::base::transpose(*a, *perm);
  } catch (std::exception& e) {
    thread_local std::string reason = e.what();
    return reason.c_str();
  }
  return nullptr;
}

template <tfcc::runtime::common::DataType DTYPE>
std::string Transpose<DTYPE>::getOperationTypeUrl() const {
  tfcc::runtime::operations::base::Transpose operation;
  return get_protobuf_type_url(operation);
}

template <tfcc::runtime::common::DataType DTYPE>
std::set<unsigned> Transpose<DTYPE>::getOperationVersions() const {
  std::set<unsigned> versions = {tfcc::runtime::operations::base::Transpose::VERSION_1};
  return versions;
}

template <tfcc::runtime::common::DataType DTYPE>
std::unique_ptr<OperationResource> Transpose<DTYPE>::process(
    const tfcc::runtime::model::Node& node, Graph& graph) const {
  if (node.inputs_size() != 1 || node.outputs_size() != 1) {
    return nullptr;
  }
  if (!node.operation().Is<tfcc::runtime::operations::base::Transpose>()) {
    return nullptr;
  }
  auto& symbolManager = getSymbolManager(graph);
  if (symbolManager.getSymbolInfo(node.outputs(0)).dtype != DTYPE ||
      symbolManager.getSymbolInfo(node.inputs(0)).dtype != DTYPE) {
    return nullptr;
  }

  std::unique_ptr<TransposeResource> resource(new TransposeResource);
  tfcc::runtime::operations::base::Transpose operation;
  node.operation().UnpackTo(&operation);
  for (unsigned v : operation.perm()) {
    resource->perm.push_back(v);
  }

  auto& jit = getJIT(graph);

  const tfcc::Tensor<T>* inputSymbol = symbolManager.getTensor(node.inputs(0), T());
  tfcc::Variable<T>* outputSymbol = symbolManager.getVariable(node.outputs(0), T());

  callFunction(jit, _wrapper_transpose<T>, inputSymbol, &resource->perm, outputSymbol);

  return std::unique_ptr<OperationResource>(std::move(resource));
}

std::vector<std::unique_ptr<Operation>> get_transpose_operations() {
#define DEFINE_FUNC(dtype) \
  operations.emplace_back(std::unique_ptr<Operation>(new Transpose<dtype>()));
  std::vector<std::unique_ptr<Operation>> operations;
  TFCC_RUNTIME_FOR_ALL_DATA_TYPE(DEFINE_FUNC);
  return operations;
}

}  // namespace base
}  // namespace runtime
}  // namespace tfcc

File: tfcc/test/python/testcase/base/transpose.py
# Copyright 2021 Wechat Group, Tencent
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import numpy as np
import test


class Transpose(test.Test):
    def getRandomTestData(self):
        dims = np.random.randint(2, 5)
        perm = np.asarray(range(dims))
        np.random.shuffle(perm)

        shape = np.random.randint(1, 10, dims)

        a = np.random.uniform(-1.0, 1.0, shape)

        expect = np.transpose(a, perm)
        return {"a": a, "perm": perm, "expect": expect}

    def getRandomTestData2(self):
        perm = np.asarray([1, 0])
        shape = np.random.randint(1, 4096, 2)

        a = np.random.uniform(-1.0, 1.0, shape)

        expect = np.transpose(a, perm)
        return {"a": a, "perm": perm, "expect": expect}

    def getRandomTestData3(self):
        perm = np.asarray([0, 2, 1])
        shape = np.random.randint(1, 128, perm.size)

        a = np.random.uniform(-1.0, 1.0, shape)

        expect = np.transpose(a, perm)
        return {"a": a, "perm": perm, "expect": expect}

    def getRandomTestData4(self):
        perm = np.asarray([1, 0, 2])
        shape = np.random.randint(1, 128, perm.size)

        a = np.random.uniform(-1.0, 1.0, shape)

        expect = np.transpose(a, perm)
        return {"a": a, "perm": perm, "expect": expect}
